crates/testing/tests/proposal_ordering.rs

```diff
@@ -0,0 +1,152 @@
+use hotshot::{tasks::task_state::CreateTaskState, types::SystemContextHandle};
+use hotshot_example_types::node_types::{MemoryImpl, TestTypes};
+use hotshot_task_impls::{consensus::ConsensusTaskState, events::HotShotEvent::*};
+use hotshot_testing::{
+    predicates::{exact, is_at_view_number, quorum_proposal_send},
+    task_helpers::vid_scheme_from_view_number,
+    view_generator::TestViewGenerator,
+};
+use hotshot_types::{data::ViewNumber, traits::node_implementation::ConsensusTime};
+use jf_primitives::vid::VidScheme;
+
+fn permute<T>(mut input: Vec<T>, permutation: Vec<usize>) -> Vec<T> {
+    let mut outputs: Vec<T> = Vec::with_capacity(input.len());
+    let mut taken = vec![false; input.len()]; // Track taken elements
+
+    for &index in &permutation {
+        // SAFETY: This is safe because we ensure that each index is accessed exactly once
+        // and `taken[index]` is set to `true` to prevent double access. This is to get around
+        // the predicate function type not being copyable below.
+        if !taken[index] {
+            unsafe {
+                let ptr = &mut input[index] as *mut T;
+                outputs.push(std::ptr::read(ptr));
+                taken[index] = true;
+            }
+        }
+    }
+
+    // Prevent a double free. Rust's borrow checker will eventually drop `input` since it is
+    // consumed by this call. However, since we pulled the memory out from underneath the data
+    // structure, we will get an error if we don't explicitly blow the memory away before it goes
+    // out of scope. This ensures that the borrow checker won't get mad at us when it finds out how
+    // we've left the place.
+    unsafe {
+        input.set_len(0);
+    }
```

Comment:
I don't think we need or want unsafe code here.  Do we need to permute the outputs?

Reply:
This appears to indeed be required, but I would be open to discussing this further. So for these inputs and outputs:
```rust
    let inputs = vec![
        QuorumProposalRecv(proposals[1].clone(), leaders[1]),
        QCFormed(either::Left(cert)),
        SendPayloadCommitmentAndMetadata(payload_commitment, (), ViewNumber::new(node_id)),
    ];
    let view_1_outputs = vec![
        exact(ViewChange(ViewNumber::new(2))),
        exact(QuorumProposalValidated(proposals[1].data.clone())),
        quorum_proposal_send(),
    ];
```
When the QuorumProposal comes last, the outputs must be in this order:
```rust
    let view_1_outputs = vec![
        quorum_proposal_send(),
        exact(ViewChange(ViewNumber::new(2))),
        exact(QuorumProposalValidated(proposals[1].data.clone())),
    ];
```
And this is the required order for any inputs where `QuorumProposalRecv` comes last. I'm trying to trace through the task, but I'll be honest that I do not know the root cause here. Is this indicative of a problem? Or am I just misunderstanding what it takes to get this to work properly?

Reply:
I think my confusion is coming from the fact that the outputs are ordered.  I was thinking they could be in any order for a stage.

Reply:
Spoke offline, opted to hard-code the two annoying cases instead of making a bespoke solution.